src/bd/primitives/Specializer.java

```diff
@@ -0,0 +1,118 @@
+package bd.primitives;
+
+import java.lang.reflect.InvocationTargetException;
+
+import com.oracle.truffle.api.dsl.NodeFactory;
+import com.oracle.truffle.api.source.SourceSection;
+
+import bd.primitives.Primitive.NoChild;
+import bd.settings.VmSettings;
+
+
+/**
+ * A Specializer defines when a node can be used as a eager primitive, how
+ * it is to be instantiated, and acts as factory for them.
+ *
+ * @param <T> the type of the nodes created by this specializer
+ * @param <Context> the type of the context object
+ * @param <ExprT> the root type of expressions used by the language
+ */
+public class Specializer<T, Context, ExprT> {
+  protected final Context                    context;
```

Comment:
VM is also such an overloaded term that it doesn't have any meaning anymore.

Yes, context is also overloaded, and it is also changing in the latest Truffle. Not really ideal. But, the objects I put in there are actually also Truffle context objects. So, it seems to be somewhat appropriate.